File: PRACTICAL_05_Transformation_CS_src/Matrix3.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace cube
{
    class Matrix3
    {// The class has nine variables, 3 rows and 3 columns
        public double A11;
        public double A12;
        public double A13;
        public double A21;
        public double A22;
        public double A23;
        public double A31;
        public double A32;
        public double A33;

        // Constructor 1 create a zero matrix
        public Matrix3()
        {
            this.A11 = 0.0;
            this.A12 = 0.0;
            this.A13 = 0.0;
            this.A21 = 0.0;
            this.A22 = 0.0;
            this.A23 = 0.0;
            this.A31 = 0.0;
            this.A32 = 0.0;
            this.A33 = 0.0;
        }

        // Constructor 2
        public Matrix3(Vector3 Row1, Vector3 Row2, Vector3 Row3)
        {  // To allow 3 rows of vectors to be declared as a matrix
            this.A11 = Row1.X;
            this.A12 = Row1.Y;
            this.A13 = Row1.Z;
            this.A21 = Row2.X;
            this.A22 = Row2.Y;
            this.A23 = Row2.Z;
            this.A31 = Row3.X;
            this.A32 = Row3.Y;
            this.A33 = Row3.Z;
        }
        // Constructor 3
        public Matrix3(  double _A11,double _A12,double _A13,
            double _A21, double _A22, double _A23,
            double _A31, double _A32,double _A33)
        {// to allow nine double values
            A11 = _A11;
            A12 = _A12;
            A13 = _A13;
            A21 = _A21;
            A22 = _A22;
            A23 = _A23;
            A31 = _A31;
            A32 = _A32;
            A33 = _A33;
        }

        public static Vector3 operator *(Matrix3 M1, Vector3 V1)
        {// An overloaded operator * to return the  product of the matrix by a vector
            return new Vector3(M1.A11 * V1.X + M1.A12 * V1.Y + M1.A13 * V1.Z,
                                M1.A21 * V1.X + M1.A22 * V1.Y + M1.A23 * V1.Z,
                                M1.A31 * V1.X + M1.A32 * V1.Y + M1.A33 * V1.Z);
        }

        public static Vector3 operator *(Vector3 V1, Matrix3 M1 )
        {// An overloaded operator * to return the  product of the matrix by a vector
            return new Vector3(M1.A11 * V1.X + M1.A21 * V1.Y + M1.A31 * V1.Z,
                                M1.A12 * V1.X + M1.A22 * V1.Y + M1.A32 * V1.Z,
                                M1.A13 * V1.X + M1.A23 * V1.Y + M1.A33 * V1.Z);
        }

        public static Matrix3 Transpose(Matrix3 M1)
        {// a method to transpose a given matrix
            return new Matrix3(M1.A11, M1.A21, M1.A31,
                M1.A12, M1.A22, M1.A32,
                M1.A13, M1.A23, M1.A33);
        }
        public static Matrix3 operator +(Matrix3 M1, Matrix3 M2)
        {// An overloaded operator + to return the  sum of two matrix 
            return new Matrix3(M1.A11 + M2.A11, M1.A12 + M2.A12, M1.A13 + M2.A13,
                M1.A21 + M2.A21, M1.A22 + M2.A22, M1.A23 + M2.A23,
                M1.A31 + M2.A31, M1.A32 + M2.A32, M1.A33 + M2.A33);
        }
        public static Matrix3 operator -(Matrix3 M1, Matrix3 M2)
        {// An overloaded operator * to return the  difference of two matrix
            return new Matrix3(M1.A11 - M2.A11, M1.A12 - M2.A12, M1.A13 - M2.A13,
                M1.A21 - M2.A21, M1.A22 - M2.A22, M1.A23 - M2.A23,
                M1.A31 - M2.A31, M1.A32 - M2.A32, M1.A33 - M2.A33);
        }

        public static Matrix3 operator *(double x, Matrix3 M1)
        {// An overloaded operator * to return the  product of the matrix by a scalar
            Matrix3 answer = new Matrix3();
            answer.A11 = M1.A11 * x ;
            answer.A12 = M1.A12 * x;
            answer.A13 = M1.A13 * x;

            answer.A21 = M1.A21 * x;
            answer.A22 = M1.A22 * x;
            answer.A23 = M1.A23 * x;

            answer.A31 = M1.A31 * x;
            answer.A32 = M1.A32 * x;
            answer.A33 = M1.A33 * x; ;
            
            return answer;
        }

        public static Matrix3 operator *(Matrix3 M1, Matrix3 M2)
        {// An overloaded operator * to return the  product of two matrix
            Matrix3 answer = new Matrix3();
            answer.A11 = M1.Row(0) * M2.Column(0);
            answer.A12 = M1.Row(0) * M2.Column(1);
            answer.A13 = M1.Row(0) * M2.Column(2);

            answer.A21 = M1.Row(1) * M2.Column(0);
            answer.A22 = M1.Row(1) * M2.Column(1);
            answer.A23 = M1.Row(1) * M2.Column(2);

            answer.A31 = M1.Row(2) * M2.Column(0);
            answer.A32 = M1.Row(2) * M2.Column(1);
            answer.A33 = M1.Row(2) * M2.Column(2);


            return answer;
        }


        public static double Determinant(Matrix3 M1)
        {// method to return the determinant of a 3x3 matrix
            //                     aei      -     ahf                  + dhc                     - gec                      +    gbh                    -     dbi   
            return M1.A11 * M1.A22 * M1.A33 - M1.A11 * M1.A32 * M1.A23 + M1.A21 * M1.A32 * M1.A13 - M1.A31 * M1.A22 * M1.A13 + M1.A31 * M1.A12 * M1.A23 - M1.A21 * M1.A12 * M1.A33;
        }

        public Vector3 Row( int i)
        { 
            // a method to return as Row as vector3 0 == first row, default == last row
            switch (i)
            {
                case 0:
                    return new Vector3(A11, A12, A13);                    
                case 1:
                    return new Vector3(A21, A22, A23);                    
                case 2:
                default:
                    return new Vector3(A31, A32, A33);                    
            }
        }

        public Vector3 Column(int i)
        {// a method to return as column as vector3 0 == first column, default == last column
            switch (i)
            {
                case 0:
                    return new Vector3(A11, A21, A31);                    
                case 1:
                    return new Vector3(A12, A22, A32);                    
                case 2:
                default:
                    return new Vector3(A13, A23, A33);                    
            }
        }


        public static Matrix3 Inverse(Matrix3 M1)
        {
            // method to return the inverse of a matrix if exists else zero vector
            double det = Determinant(M1);
            if (det == 0)
                return new Matrix3();
            else
            {
                double scale = 1 / det;
                Matrix3 answer = new Matrix3();
                answer.A11 = scale * (M1.A22 * M1.A33 - M1.A23 * M1.A32); // ei-fh
                answer.A12 = scale * (M1.A13 * M1.A32 - M1.A12 * M1.A33); // ch-bi
                answer.A13 = scale * (M1.A12 * M1.A23 - M1.A13 * M1.A22); // ch-bi

                answer.A21 = scale * (M1.A23 * M1.A31 - M1.A21 * M1.A33); // fg-di
                answer.A22 = scale * (M1.A11 * M1.A33 - M1.A13 * M1.A31); // ai-cg
                answer.A23 = scale * (M1.A13 * M1.A21 - M1.A11 * M1.A23); // cd-af


                answer.A31 = scale * (M1.A21 * M1.A32 - M1.A22 * M1.A31); // dh-eg
                answer.A32 = scale * (M1.A12 * M1.A31 - M1.A11 * M1.A32); // bg-ah
                answer.A33 = scale * (M1.A11 * M1.A22 - M1.A12 * M1.A21); // ae-bd

                return answer;
            }

        }

        public static Matrix3 Rotation(int _angle)
        {
            Double radians = Math.PI / 180 * _angle;
            Matrix3 answer = new Matrix3();
            answer.A11 = Math.Cos(radians);
            answer.A12 = Math.Sin(radians);
            answer.A13 = 0;
            answer.A21 = -Math.Sin(radians);
            answer.A22 = Math.Cos(radians);
            answer.A23 = 0;
            answer.A31 = 0;
            answer.A32 = 0;
            answer.A33 = 1;

            return answer;
        }

        
        public static Matrix3 Translate(int dx, int dy)
        {
            Matrix3 answer = new Matrix3();
            answer.A11 = 1;
            answer.A12 = 0;
            answer.A13 = 0;
            answer.A21 = 0;
            answer.A22 = 1;
            answer.A23 = 0;
            answer.A31 = dx;
            answer.A32 = dy;
            answer.A33 = 1;

            return answer;
        }

        public static Matrix3 Scale(int dx, int dy)
        {
            Matrix3 answer = new Matrix3();
            answer.A11 = (double)dx/100;
            answer.A12 = 0;
            answer.A13 = 0;
            answer.A21 = 0;
            answer.A22 = (double)dy/100;
            answer.A23 = 0;
            answer.A31 = 0;
            answer.A32 = 0;
            answer.A33 = 1;

            return answer;
        }

        public static Matrix3 operator -(Matrix3 M1)
        {
            return -1 * M1;
        }

        public static Matrix3 RotationX(int _angle)
        {
            Double radians = Math.PI / 180 * _angle;
            Matrix3 answer = new Matrix3();
            answer.A11 = 1;
            answer.A12 = 0;
            answer.A13 = 0;
            answer.A21 = 0;
            answer.A22 = Math.Cos(radians);
            answer.A23 = -Math.Sin(radians);
            answer.A31 = 0;
            answer.A32 = Math.Sin(radians);
            answer.A33 = Math.Cos(radians);

            return answer;
        }
        public static Matrix3 RotationY(int _angle)
        {
            Double radians = Math.PI / 180 * _angle;
            Matrix3 answer = new Matrix3();
            answer.A11 = Math.Cos(radians);
            answer.A12 = 0;
            answer.A13 = Math.Sin(radians);
            answer.A21 = 0;
            answer.A22 = 1;
            answer.A23 = 0;
            answer.A31 = -Math.Sin(radians);
            answer.A32 = 0;
            answer.A33 = Math.Cos(radians);

            return answer;
        }
        public static Matrix3 RotationZ(int _angle)
        {
            Double radians = Math.PI / 180 * _angle;
            Matrix3 answer = new Matrix3();
            answer.A11 = Math.Cos(radians);
            answer.A12 = -Math.Sin(radians);
            answer.A13 = 0;
            answer.A21 = Math.Sin(radians);
            answer.A22 = Math.Cos(radians);
            answer.A23 = 0;
            answer.A31 = 0;
            answer.A32 = 0;
            answer.A33 = 1;

            return answer;
        }

        public static Matrix3 Scale3D(int dx)
        {
            Matrix3 answer = new Matrix3();
            answer.A11 = (double)dx / 100;
            answer.A12 = 0;
            answer.A13 = 0;
            answer.A21 = 0;
            answer.A22 = (double)dx / 100;
            answer.A23 = 0;
            answer.A31 = 0;
            answer.A32 = 0;
            answer.A33 = (double)dx / 100;

            return answer;
        }
    }
}

File: PRACTICAL_05_Transformation_CS_src/MatrixQuatTest/OpenGL Lab04/Matrix3.cpp
#include "Matrix3.h"
/// <summary>
/// default constructor
/// </summary>
Matrix3::Matrix3() :
	m_a11(0.0),
	m_a12(0.0),
	m_a13(0.0),
	m_a21(0.0),
	m_a22(0.0),
	m_a23(0.0),
	m_a31(0.0),
	m_a32(0.0),
	m_a33(0.0)
{

}
/// <summary>
/// Overloaded constructor that takes in 3 vectors
/// </summary>
/// <param name="row1">first row</param>
/// <param name="row2">second row</param>
/// <param name="row3">third row</param>
Matrix3::Matrix3(Vector3 row1, Vector3 row2, Vector3 row3) :
	m_a11(row1.getX()),
	m_a12(row1.getY()),
	m_a13(row1.getZ()),
	m_a21(row2.getX()),
	m_a22(row2.getY()),
	m_a23(row2.getZ()),
	m_a31(row3.getX()),
	m_a32(row3.getY()),
	m_a33(row3.getZ())
{
}

/// <summary>
/// Overloaded constructor that takes in individual values
/// </summary>
/// <param name="a11">11</param>
/// <param name="a12">12</param>
/// <param name="a13">12</param>
/// <param name="a21">21</param>
/// <param name="a22">22</param>
/// <param name="a23">23</param>
/// <param name="a31">31</param>
/// <param name="a32">32</param>
/// <param name="a33">33</param>
Matrix3::Matrix3(double a11, double a12, double a13, 
				 double a21, double a22, double a23,
				 double a31, double a32, double a33) :
	m_a11(a11),
	m_a12(a12),
	m_a13(a13),
	m_a21(a21),
	m_a22(a22),
	m_a23(a23),
	m_a31(a31),
	m_a32(a32),
	m_a33(a33)
{
}
/// <summary>
/// Overloaded operator that multiplies a matrix by a vector
/// </summary>
/// <param name="V">the vector to multiply by</param>
/// <returns>produckt of vector by matrix</returns>
Vector3 Matrix3::operator*(const Vector3 & V) const
{
	return Vector3(m_a11 * V.getX() + m_a12 * V.getY() + m_a13 * V.getZ(),
				   m_a21 * V.getX() + m_a22 * V.getY() + m_a23 * V.getZ(),
				   m_a31 * V.getX() + m_a32 * V.getY() + m_a33 * V.getZ());
}
/// <summary>
/// Overloaded operator to allow addition of two matrices
/// </summary>
/// <param name="m1">matrix m1</param>
/// <returns>the result of adding two matrices</returns>
Matrix3 Matrix3::operator+(const Matrix3 &m1) const
{
	return Matrix3( Row(0) + m1.Row(0),
					Row(1) + m1.Row(1),
					Row(2) + m1.Row(2));
}
/// <summary>
/// Overload operator to allow the subtraction of one matrix from another
/// </summary>
/// <param name="m1">matrix m1</param>
/// <returns>result of subtracting two matrices</returns>
Matrix3 Matrix3::operator-(const Matrix3 &m1) const
{
	return Matrix3( Row(0) - m1.Row(0),
					Row(1) - m1.Row(1),
					Row(2) - m1.Row(2));
}
/// <summary>
/// Allows to multiply a matrix by a number
/// </summary>
/// <param name="x">number to multiply by</param>
/// <returns>result of matrix multiplied by a number</returns>
Matrix3 Matrix3::operator*(const double& x) const
{
	return Matrix3( Row(0) * x, Row(1) * x, Row(2) * x);
}
/// <summary>
/// Returns a Row of a matrix as a vector3 0 being first row and 2/default the last
/// </summary>
/// <param name="">number of row</param>
/// <returns>row of matrix as vector3</returns>
Vector3 Matrix3::Row(int i) const
{
	switch (i)
	{
		case 0:
			return Vector3(m_a11, m_a12, m_a13);
		case 1:
			return Vector3(m_a21, m_a22, m_a23);
		case 2:
		default:
			return Vector3(m_a31, m_a32, m_a33);
	}
}
/// <summary>
/// Returns a Column of a matrix as a vector3 0 being first col and 2/default the last
/// </summary>
/// <param name="i">the column num</param>
/// <returns>column of matrix as vector3</returns>
Vector3 Matrix3::Column(int i) const
{
	switch (i)
	{
		case 0:
			return Vector3(m_a11, m_a21, m_a31);
		case 1:
			return Vector3(m_a12, m_a22, m_a32);
		case 2:
		default:
			return Vector3(m_a13, m_a23, m_a33);
	}
}
/// <summary>
/// Transposes the matrix turning each row into a column
/// </summary>
/// <returns>Transposed matrix</returns>
Matrix3 Matrix3::transpose(const Matrix3& M)
{
	return Matrix3( M.Column(0),
					M.Column(1),
					M.Column(2));
}
/// <summary>
/// Returns the determinant of a 3x3 matrix
/// (11)(22)(33) - (11)(32)(23) + (21)(32)(13) - (31)(22)(13) + (31)(12)(23) - (21)(12)(33)   
/// </summary>
/// <returns>the determinant</returns>
double Matrix3::determinant() const
{
	return m_a11 * m_a22 * m_a33 - m_a11 * m_a32 * m_a23 + m_a21 * m_a32 * m_a13 - m_a31 * m_a22 * m_a13 + m_a31 * m_a12 * m_a23 - m_a21 * m_a12 * m_a33;
}
/// <summary>
/// Returns an inverse of a matrix if one exists else default matrix
/// </summary>
/// <returns>inverse of matrix</returns>
Matrix3 Matrix3::inverse(const Matrix3& M)
{
	double det = M.determinant();
	if (det == 0)
		return Matrix3();
	else
	{
		double scale = 1 / det;
		return Matrix3(M.m_a22 * M.m_a33 - M.m_a23 * M.m_a32, M.m_a13 * M.m_a32 - M.m_a12 * M.m_a33, M.m_a12 * M.m_a23 - M.m_a13 * M.m_a22,
					   M.m_a23 * M.m_a31 - M.m_a21 * M.m_a33, M.m_a11 * M.m_a33 - M.m_a13 * M.m_a31, M.m_a13 * M.m_a21 - M.m_a11 * M.m_a23,
					   M.m_a21 * M.m_a32 - M.m_a22 * M.m_a31, M.m_a12 * M.m_a31 - M.m_a11 * M.m_a32, M.m_a11 * M.m_a22 - M.m_a12 * M.m_a21) * scale;
	}
}
/// <summary>
/// Rotates the matrix about the input axis
/// </summary>
/// <param name="axis">axis about which we rotate</param>
/// <param name="angle">angle of rotation</param>
/// <returns>rotation matrix</returns>
Matrix3 Matrix3::rotation(const Axis &axis, const int &angle)
{
	double radians = acos(-1) / 180 * angle;
	switch (axis)
	{
	case Axis::X:
		return Matrix3( 1,		0,				0,
						0, cos(radians), -sin(radians),
						0, sin(radians), cos(radians));
	case Axis::Y:
		return Matrix3( cos(radians),0,sin(radians),
						0,			1,		0,
						-sin(radians),0,cos(radians));
	case Axis::Z:
		return Matrix3(	cos(radians),-sin(radians),0,
						sin(radians),cos(radians), 0,
						0,				0,		   1);
	default:
		return Matrix3();
	}
}
/// <summary>
/// The creation of translation matrix
/// </summary>
/// <param name="dx">translation in x</param>
/// <param name="dy">translation in y</param>
/// <returns>the translation matrix</returns>
Matrix3 Matrix3::translate(const int &dx, const int &dy)
{
	return Matrix3( 1,0,dx,
					0,1,dy,
					0,0,1);
}
/// <summary>
/// Returns the scale matrix
/// </summary>
/// <param name="dx">scale in x</param>
/// <param name="dy">scale in y</param>
/// <param name="dz">scale in z</param>
/// <returns>The scale matrix</returns>
Matrix3 Matrix3::scale(const double &dx, const double &dy, const double &dz)
{
	return Matrix3( dx,0,0,
					0,dy,0,
					0,0,dz);
}

